contracts/modules/Experimental/TransferManager/LockupVolumeRestrictionTM.sol

```diff
@@ -229,164 +238,208 @@ contract LockupVolumeRestrictionTM is ITransferManager {
-                }
-
-            }
-            return Result.VALID;
-        }
+    function _addLockUp(
+        address _userAddress,
+        uint256 _lockupAmount,
+        uint256 _startTime,
+        uint256 _lockUpPeriodSeconds,
+        uint256 _releaseFrequencySeconds
+    ) 
+        internal 
+    {   
+        /*solium-disable-next-line security/no-block-members*/
+        require(_startTime >= now, "start time is in past");
+        _checkLockUpParams(_userAddress, _lockupAmount, _lockUpPeriodSeconds, _releaseFrequencySeconds);
+
+        lockUps[_userAddress].push(LockUp(_lockupAmount, _startTime, _lockUpPeriodSeconds, _releaseFrequencySeconds));
 
-        return _checkIfUnlockedTokenTransferIsPossible(userAddress, amount, tokenSums[1], tokenSums[2]);
+        emit AddNewLockUp(
+            _userAddress,
+            _lockupAmount,
+            _startTime,
+            _lockUpPeriodSeconds,
+            _releaseFrequencySeconds,
+            lockUps[_userAddress].length -1
+        );
     }
 
-    function _checkIfUnlockedTokenTransferIsPossible(
-        address userAddress,
-        uint amount,
-        uint totalSum,
-        uint alreadyWithdrawnSum
-        ) internal view returns (Result) {
-        // the amount the user wants to withdraw is greater than their allowed amounts according to the lockups.  however, if the user has like, 10 tokens, but only 4 are locked up, we should let the transfer go through for those 6 that aren't locked up
-        uint currentUserBalance = ISecurityToken(securityToken).balanceOf(userAddress);
-        uint stillLockedAmount = totalSum.sub(alreadyWithdrawnSum);
-        if (currentUserBalance >= stillLockedAmount && amount <= currentUserBalance.sub(stillLockedAmount)) {
-            // the user has more tokens in their balance than are actually locked up.  they should be allowed to withdraw the difference
-            return Result.VALID;
+    /**
+     * @notice Checks whether the transfer is allowed
+     * @param _userAddress Address of the user whose lock ups should be checked
+     * @param _amount Amount of tokens that need to transact
+     */
+    function _checkIfValidTransfer(address _userAddress, uint256 _amount) internal view returns (Result) {
+        uint256 totalRemainingLockedAmount = getLockedTokenToUser(_userAddress);
+        // Present balance of the user
+        uint256 currentBalance = ISecurityToken(securityToken).balanceOf(_userAddress);
+        if ((currentBalance.sub(_amount)) >= totalRemainingLockedAmount) {
+            return Result.NA;
         }
         return Result.INVALID;
     }
 
+    /**
+     * @notice Provide the unlock amount for the given lockup for a particular user
+     */
+    function _getUnlockedAmountForLockup(LockUp[] userLockup, uint256 _lockupIndex) internal view returns (uint256) {
+            /*solium-disable-next-line security/no-block-members*/
+            if (userLockup[_lockupIndex].startTime > now) {
```

Comment:
Maybe we should also be checking if:
`(userLockup[_lockupIndex].startTime.add(userLockup[_lockupIndex].lockUpPeriodSeconds) <= now)`
as an optimisation?